runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -193,36 +193,109 @@ protected boolean discoverAddressSpace(final UUID streamId,
+        // addresses (for instance for the first checkpoint) might have been already
+        // trimmed from the log, but stream maps (sequencer) still do not reflect it
+        // (causing unnecessary TrimmedExceptions--as this checkpoint is not even
+        // needed in the first place).
+        Iterable<List<Long>> batches = Iterables.partition(checkpointAddresses,
+                runtime.getParameters().getCheckpointReadBatchSize());
+
+        boolean checkpointResolved = false;
+        for (List<Long> batch : batches) {
+            if (checkpointResolved) {
+                // No need to fetch next batch of addresses, checkpoint has been resolved.
+                break;
+            }
+
+            try {
+                List<ILogData> entries = readAll(batch);
+                for (ILogData data : entries) {
+                    checkpointResolved = filterCheckpointEntry(data, filter, queue);
+                    if (checkpointResolved) {
+                        break;
+                    }
+                }
+            } catch (TrimmedException te) {
+                // Read one entry at a time for the last failed batch, this way we might load
+                // the required checkpoint entries until the stop condition is fulfilled
+                // without hitting a trimmed position. If we do hit a trim it is an actual
+                // TrimmedException
+                checkpointResolved = processCheckpointBatchByEntry(batch, filter, queue);
+            }
+        }
+    }
 
-        // Read all checkpoint entries in a batch read
+    /**
+     * Process a batch of addresses as single entries.
+     *
+     * @param batch list of addresses to read.
+     * @param filter filter to apply to checkpoint data.
+     * @param queue read queue.
+     *
+     * @return True, resolved checkpoint (reached end of valid checkpoint).
+     *         False, otherwise.
+     */
+    private boolean processCheckpointBatchByEntry(List<Long> batch,
+                                                  Function<ILogData, BackpointerOp> filter,
+                                                  NavigableSet<Long> queue) {
+        long lastReadAddress = Address.NON_ADDRESS;
         try {
-            checkpointData = readAll(checkpointAddresses);
-        } catch (TrimmedException te) {
+            boolean checkpointResolved;
+            for (long address : batch) {
+                lastReadAddress = address;
+                ILogData data = read(address);
+                checkpointResolved = filterCheckpointEntry(data, filter, queue);
+                if (checkpointResolved) {
+                    // Return if checkpoint has already been resolved (reached stop).
+                    return true;
+                }
+            }
+        } catch (TrimmedException ste) {
+            // The valid checkpoint has been trimmed.
             if (options.ignoreTrimmed) {
```

Comment:
I think we should put this ignoreTrimmed option check into `processTrimmedException`, for two reasons: 1) cleaner code organization 2) `processTrimmedException` is basically checking if this TrimmedException is retryable. If we ignore trim, there is no point doing this retryable check, right?